### PR TITLE
ISSUE-414 Implement Forward/get JMAP method

### PR DESCRIPTION
**Pull request by Arsnael**

### CODE REVIEW

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/model/ForwardGet.scala [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+case class ForwardGetRequest(accountId: AccountId,
+                             ids: Option[ForwardIds],
+                             properties: Option[Properties]) extends WithAccountId
+
+case class ForwardNotFound(value: Set[UnparsedForwardId]) {
+  def merge(other: ForwardNotFound): ForwardNotFound = ForwardNotFound(this.value ++ other.value)
+}
+
+case class ForwardId()
```
Comment (chibenwa) [inline tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/model/ForwardGet.scala:21]: case object?

Comment (Arsnael) [reply tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/model/ForwardGet.scala:21]: Hmm rationals behind that? I mostly sticked to something similar to VacationResponseGetMethod as it's a simple singleton too. 

`ForwardId` seems used like for the serialization for example. Is case object justified? (maybe it is, I'm just not sure :))

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/ForwardGetMethod.scala [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+        .flatMap(id => id match {
+          case UNPARSED_SINGLETON => getForwardsSingleton(mailboxSession)
+            .map(ForwardGetResult.found)
+          case _ => SMono.just(ForwardGetResult.notFound(id))
+        })
+    }
+
+  private def getForwardsSingleton(mailboxSession: MailboxSession): SMono[Forwards] = {
+    val userMailAddress: MailAddress = mailboxSession.getUser.asMailAddress
+    SMono.just(recipientRewriteTable.getStoredMappings(MappingSource.fromMailAddress(userMailAddress))
```
Comment (Arsnael) [inline tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/ForwardGetMethod.scala:128]: It's green for the IT tests with memory, works as expected.

It's not for Distributed and I'm struggling.

When getting the mappings from `CassandraRecipientRewriteTable`, it does a `.blockOptional()` in the middle and reactor does not like it when I wrap it in my method processing... I tried to move the mappings out of the Mono but still get the same issue when calling `getStoredMappings`

The trace:
```
16:03:19.148 [ERROR] o.a.j.j.c.ProblemDetails - Unexpected error upon API request
java.lang.IllegalStateException: blockOptional() is blocking, which is not supported in thread reactor-http-epoll-2
	at reactor.core.publisher.BlockingOptionalMonoSubscriber.blockingGet(BlockingOptionalMonoSubscriber.java:104)
	at reactor.core.publisher.Mono.blockOptional(Mono.java:1752)
	at org.apache.james.rrt.cassandra.CassandraRecipientRewriteTable.getStoredMappings(CassandraRecipientRewriteTable.java:65)
	at com.linagora.tmail.james.jmap.method.ForwardGetMethod.getForwardsSingleton(ForwardGetMethod.scala:128)
	at com.linagora.tmail.james.jmap.method.ForwardGetMethod.getForwards(ForwardGetMethod.scala:115)
	at com.linagora.tmail.james.jmap.method.ForwardGetMethod.doProcess(ForwardGetMethod.scala:87)
	at com.linagora.tmail.james.jmap.method.ForwardGetMethod.doProcess(ForwardGetMethod.scala:77)
	at org.apache.james.jmap.method.MethodRequiringAccountId.$anonfun$process$2(Method.scala:61)
	at scala.util.Either.map(Either.scala:382)
	at org.apache.james.jmap.method.MethodRequiringAccountId.$anonfun$process$1(Method.scala:59)
	at scala.util.Either.flatMap(Either.scala:352)
	at org.apache.james.jmap.method.MethodRequiringAccountId.process(Method.scala:58)
	at org.apache.james.jmap.method.MethodRequiringAccountId.process$(Method.scala:56)
	at com.linagora.tmail.james.jmap.method.ForwardGetMethod.process(ForwardGetMethod.scala:77)
	at org.apache.james.jmap.routes.JMAPApi.$anonfun$processMethodWithMatchName$3(JmapApi.scala:89)
	at scala.util.Either.fold(Either.scala:189)
	at org.apache.james.jmap.routes.JMAPApi.$anonfun$processMethodWithMatchName$1(JmapApi.scala:89)
	at scala.Option.map(Option.scala:242)
	at org.apache.james.jmap.routes.JMAPApi.processMethodWithMatchName(JmapApi.scala:87)
	at org.apache.james.jmap.routes.JMAPApi.process(JmapApi.scala:81)
	at org.apache.james.jmap.routes.JMAPApi.$anonfun$processSequentiallyAndUpdateContext$5(JmapApi.scala:66)
	at reactor.core.scala.publisher.package$.$anonfun$scalaFunction2JavaFunction$1(package.scala:51)
	at reactor.core.publisher.FluxFlatMap.trySubscribeScalarMap(FluxFlatMap.java:152)
	at reactor.core.publisher.MonoFlatMapMany.subscribeOrReturn(MonoFlatMapMany.java:49)
	at reactor.core.publisher.FluxFromMonoOperator.subscribe(FluxFromMonoOperator.java:76)
	at reactor.core.publisher.FluxReplay.connect(FluxReplay.java:1120)
	at reactor.core.publisher.FluxAutoConnectFuseable.subscribe(FluxAutoConnectFuseable.java:63)
	at reactor.core.publisher.Flux.subscribe(Flux.java:8469)
	at reactor.core.scala.publisher.SFlux.subscribe(SFlux.scala:909)
	at reactor.core.scala.publisher.SFlux.subscribe$(SFlux.scala:909)
	at reactor.core.scala.publisher.ReactiveSFlux.subscribe(ReactiveSFlux.scala:8)
	at reactor.core.publisher.FluxConcatMap$ConcatMapImmediate.drain(FluxConcatMap.java:451)
	at reactor.core.publisher.FluxConcatMap$ConcatMapImmediate.onSubscribe(FluxConcatMap.java:219)
	at reactor.core.publisher.FluxIterable.subscribe(FluxIterable.java:165)
	at reactor.core.publisher.FluxIterable.subscribe(FluxIterable.java:87)
	at reactor.core.publisher.InternalMonoOperator.subscribe(InternalMonoOperator.java:64)
	at reactor.core.publisher.MonoFlatMap$FlatMapMain.onNext(MonoFlatMap.java:157)
	at reactor.core.publisher.FluxMapFuseable$MapFuseableSubscriber.onNext(FluxMapFuseable.java:127)
	at reactor.core.publisher.Operators$MonoSubscriber.complete(Operators.java:1816)
	at reactor.core.publisher.MonoReduceSeed$ReduceSeedSubscriber.onComplete(MonoReduceSeed.java:165)
	at reactor.core.publisher.FluxIterable$IterableSubscription.fastPath(FluxIterable.java:362)
	at reactor.core.publisher.FluxIterable$IterableSubscription.request(FluxIterable.java:227)
	at reactor.core.publisher.MonoReduceSeed$ReduceSeedSubscriber.onSubscribe(MonoReduceSeed.java:113)
	at reactor.core.publisher.FluxIterable.subscribe(FluxIterable.java:165)
	at reactor.core.publisher.FluxIterable.subscribe(FluxIterable.java:87)
	at reactor.core.publisher.InternalMonoOperator.subscribe(InternalMonoOperator.java:64)
	at reactor.core.publisher.MonoFlatMap$FlatMapMain.onNext(MonoFlatMap.java:157)
	at reactor.core.publisher.FluxHandle$HandleSubscriber.onNext(FluxHandle.java:119)
	at reactor.core.publisher.FluxHandle$HandleSubscriber.onNext(FluxHandle.java:119)
	at reactor.core.publisher.FluxMap$MapConditionalSubscriber.onNext(FluxMap.java:220)
	at reactor.core.publisher.FluxDoFinally$DoFinallySubscriber.onNext(FluxDoFinally.java:130)
	at reactor.core.publisher.FluxHandleFuseable$HandleFuseableSubscriber.onNext(FluxHandleFuseable.java:184)
	at reactor.core.publisher.FluxContextWrite$ContextWriteSubscriber.onNext(FluxContextWrite.java:107)
	at reactor.core.publisher.Operators$MonoSubscriber.complete(Operators.java:1816)
	at reactor.core.publisher.MonoCollectList$MonoCollectListSubscriber.onComplete(MonoCollectList.java:128)
	at reactor.core.publisher.FluxPeek$PeekSubscriber.onComplete(FluxPeek.java:260)
	at reactor.core.publisher.FluxMap$MapSubscriber.onComplete(FluxMap.java:142)
	at reactor.netty.channel.FluxReceive.terminateReceiver(FluxReceive.java:468)
	at reactor.netty.channel.FluxReceive.drainReceiver(FluxReceive.java:260)
	at reactor.netty.channel.FluxReceive.request(FluxReceive.java:129)
	at reactor.core.publisher.FluxMap$MapSubscriber.request(FluxMap.java:162)
	at reactor.core.publisher.FluxPeek$PeekSubscriber.request(FluxPeek.java:138)
	at reactor.core.publisher.MonoCollectList$MonoCollectListSubscriber.onSubscribe(MonoCollectList.java:79)
	at reactor.core.publisher.FluxPeek$PeekSubscriber.onSubscribe(FluxPeek.java:171)
	at reactor.core.publisher.FluxMap$MapSubscriber.onSubscribe(FluxMap.java:92)
	at reactor.netty.channel.FluxReceive.startReceiver(FluxReceive.java:167)
	at reactor.netty.channel.FluxReceive.lambda$subscribe$2(FluxReceive.java:146)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:164)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:469)
	at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:384)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:986)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
```

Trying to investigate but not really sure what's going on there... or do we need to reactify that API?

Comment (Arsnael) [reply tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/ForwardGetMethod.scala:128]: Nevermind... :)

File: tmail-backend/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/tmail/james/common/LinagoraForwardGetMethodContract.scala [hunk 3]
```diff
@@ -0,0 +1,855 @@
...
+         |  "sessionState": "${SESSION_STATE.value}",
+         |  "methodResponses": [[
+         |    "Forward/get",
+         |    {
+         |      "accountId": "29883977c13473ae7cb7678ef767cbfbaffc8a44a6e463d971d23a65c1dc4af6",
+         |      "state": "${INSTANCE.value}",
+         |      "list": [
+         |        {
+         |          "id":"singleton",
+         |          "localCopy": false,
```
Comment (chibenwa) [inline tmail-backend/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/tmail/james/common/LinagoraForwardGetMethodContract.scala:77]: IMO by default when there's no forwars we should say "local copy" -> true as the guy will receive the mail.

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/model/ForwardGet.scala (hunk 1 above)
Comment (chibenwa) [reply tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/model/ForwardGet.scala:21]: > . Is case object justified? (maybe it is, I'm just not sure :))

Yes that will empower us to not create tons of ForwardId instances...

File: tmail-backend/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/tmail/james/common/LinagoraForwardGetMethodContract.scala (hunk 3 above)
Comment (Arsnael) [reply tmail-backend/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/tmail/james/common/LinagoraForwardGetMethodContract.scala:77]: The local copy radio button seemed completely independent of that on esn, thus I kept the behavior. But your remark makes sense and I don't mind. Will check

**Review by chibenwa — APPROVED** [review]

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/json/ForwardSerializer.scala [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+import com.linagora.tmail.james.jmap.model.Forwards.FORWARD_ID
+import com.linagora.tmail.james.jmap.model.{Forward, ForwardGetRequest, ForwardGetResponse, ForwardId, ForwardIds, ForwardNotFound, Forwards, LocalCopy, UnparsedForwardId}
+import eu.timepit.refined
+import org.apache.james.core.MailAddress
+import org.apache.james.jmap.core.Id.IdConstraint
+import org.apache.james.jmap.core.{Properties, UuidState}
+import play.api.libs.json._
+
+class ForwardSerializer {
```
Comment (vttranlina) [inline tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/json/ForwardSerializer.scala:10]: It can be `object`

**Comment by quantranhong1999** [comment]
Read it.

**Pull request merged**